java/org/contikios/cooja/util/CmdUtils.java

```diff
@@ -0,0 +1,108 @@
+ *
+ */
+
+package org.contikios.cooja.util;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import org.contikios.cooja.Cooja;
+import org.contikios.cooja.dialogs.MessageContainer;
+import org.contikios.cooja.dialogs.MessageList;
+
+/**
+ * Utility functions for command execution.
+ */
+public class CmdUtils {
+  /**
+   * Executes command and returns the output.
+   *
+   * @param command Command to execute.
+   * @param gui Cooja object.
+   * @param synchronous Wait for the output readers to terminate before returning.
+   * @return Command execution output
+   */
+  public static String[] run(String command, Cooja gui, boolean synchronous) throws Exception {
+    command = gui.resolveConfigDir(Cooja.resolvePathIdentifiers(command));
+
+    ArrayList<String> output = new ArrayList<>();
+    final MessageList cmdIO = MessageContainer.createMessageList(Cooja.isVisualized());
+    try {
+      var pb = new ProcessBuilder("/bin/sh", "-c", command);
+      pb.directory(new File(gui.getConfigDir()));
+      final Process p = pb.start();
+      var stderr = new Thread(() -> {
+        try (var errorInput = new BufferedReader(new InputStreamReader(p.getErrorStream(), UTF_8))) {
+          String line;
+          while ((line = errorInput.readLine()) != null) {
+            cmdIO.addMessage(line, MessageList.ERROR);
+          }
+        } catch (IOException e) {
+          cmdIO.addMessage("Error reading from stderr: " + e.getMessage(), MessageList.ERROR);
+        }
+      });
+      stderr.setDaemon(true);
+      stderr.start();
+      var stdout = new Thread(() -> {
+        try (var input = new BufferedReader(new InputStreamReader(p.getInputStream(), UTF_8))) {
+          String line;
+          while ((line = input.readLine()) != null) {
+            cmdIO.addMessage(line, MessageList.NORMAL);
+            output.add(line);
+          }
+        } catch (IOException e) {
+          cmdIO.addMessage("Error reading from stdout: " + e.getMessage(), MessageList.ERROR);
+        }
+      });
+      stdout.setDaemon(true);
+      stdout.start();
+
+      int ret = p.waitFor();
+      if (synchronous) {
+        stdout.join();
+        stderr.join();
+      }
+      if (ret != 0) {
+        throw createException("Command failed with error: " + ret, null, command, cmdIO);
+      }
+      return output.toArray(new String[0]);
+    } catch (InterruptedException | IOException err) {
+      throw createException("Command error: " + err.getMessage(), err, command, cmdIO);
+    }
+  }
+
+  private static Exception createException(String message, Throwable err,
+                                           String command, MessageList outputList) {
+    outputList.addMessage("Failed to run command: " + command, MessageList.ERROR);
+    return new Exception(message, err);
```

Comment:
Some exceptions such as `MoteTypeCreateException` can include a MessageList that will be shown in the error dialog when running Cooja with GUI. `MoteTypeCreateException` might not be suitable for a generic command executer but perhaps a new type of exception (or interface) that can include a MessageList? But this is something that might be better addressed in another PR.

Reply:
Yes. Some parts of the code that run things in Cooja should be removed, other code should probably be migrated to this helper. With more callers, it will be more clear what is suitable for this method.